src/model/chart-model.ts

```diff
@@ -467,22 +467,30 @@ export class ChartModel implements IDestroyable {
 	}
 
 	public updateTimeScale(newBaseIndex: TimePointIndex, newPoints?: readonly TimeScalePoint[]): void {
+		const oldFirstTime = this._timeScale.indexToTime(0 as TimePointIndex);
+
 		if (newPoints !== undefined) {
 			this._timeScale.update(newPoints);
 		}
 
+		const newFirstTime = this._timeScale.indexToTime(0 as TimePointIndex);
+
 		const currentBaseIndex = this._timeScale.baseIndex();
 		const visibleBars = this._timeScale.visibleStrictRange();
 
 		// if time scale cannot return current visible bars range (e.g. time scale has zero-width)
 		// then we do not need to update right offset to shift visible bars range to have the same right offset as we have before new bar
 		// (and actually we cannot)
-		if (visibleBars !== null) {
+		if (visibleBars !== null && oldFirstTime !== null && newFirstTime !== null) {
 			const isLastSeriesBarVisible = visibleBars.contains(currentBaseIndex);
+			const isFirstSeriesBarVisible = visibleBars.contains(0 as TimePointIndex);
+			const isFirstTimeChanged = oldFirstTime.timestamp > newFirstTime.timestamp;
+			const isDataAdded = newBaseIndex > currentBaseIndex;
```

Comment:
I guess that `newBaseIndex > currentBaseIndex` doesn't really mean that "data was added". base index - it's a term to show the latest _real_ bar across all series on the chart (because of whitespaces, where timescale might have data additional data from the right, but no series has data there).

Thus, I think it's something like "added new bars from the right" or so (but actually this also doesn't mean that the data was added to the right though...)

Reply:
Or maybe we can just "inline" it to the `if` statement below (like it was previously)?..

Reply:
Ok. I can also suggest "inline" it to the `isDataAddedToRight`
```suggestion
			const isDataAddedToRight = newBaseIndex > currentBaseIndex && !isDataAddedToLeft;
```

Reply:
Why it's so `&& !isDataAddedToLeft`? 🙂 I think that we can add from both side at the same time.